middle_end/flambda/types/type_of_kind/type_of_kind_naked_immediate0.rec.ml

```diff
@@ -78,113 +78,187 @@ module Make_meet_or_join
     with type typing_env := Typing_env.t
     with type typing_env_extension := Typing_env_extension.t) =
 struct
-  let bad_meet_or_join env t1 t2 =
-    Misc.fatal_errorf "Bad naked immediate meet/join with %a and %a in env:@ %a"
-      print t1
-      print t2
-      TE.print (Meet_or_join_env.target_join_env env)
-
-  let meet_or_join env t1 t2 : _ Or_bottom_or_absorbing.t =
-  (*
-    Format.eprintf "NN meet_or_join %a and %a\n%!"
-      print t1 print t2 (*Typing_env.print (Meet_env.env env)*);
-      *)
+  let meet env t1 t2 : _ Or_bottom.t =
     match t1, t2 with
     | Naked_immediates is1, Naked_immediates is2 ->
-      let is = E.Target_imm.Set.union_or_inter is1 is2 in
+      let is = I.Set.inter is1 is2 in
       if I.Set.is_empty is then Bottom
       else Ok (Naked_immediates is, TEE.empty ())
     | Is_int ty1, Is_int ty2 ->
-      Or_bottom_or_absorbing.of_or_bottom (E.switch T.meet T.join' env ty1 ty2)
+      Or_bottom.map (T.meet env ty1 ty2)
         ~f:(fun (ty, env_extension) -> Is_int ty, env_extension)
     | Get_tag ty1, Get_tag ty2 ->
-      Or_bottom_or_absorbing.of_or_bottom (E.switch T.meet T.join' env ty1 ty2)
+      Or_bottom.map (T.meet env ty1 ty2)
         ~f:(fun (ty, env_extension) -> Get_tag ty, env_extension)
     | Is_int ty, Naked_immediates is_int ->
       begin match I.Set.elements is_int with
       | [] -> Bottom
       | [is_int] ->
         let shape =
-          if I.equal is_int I.zero then T.any_block ()
-          else if I.equal is_int I.one then T.any_tagged_immediate ()
-          else bad_meet_or_join env t1 t2
+          if I.equal is_int I.zero then Some (T.any_block ())
+          else if I.equal is_int I.one then Some (T.any_tagged_immediate ())
+          else None
         in
-        Or_bottom_or_absorbing.of_or_bottom
-          (E.switch T.meet T.join' env ty shape)
-          ~f:(fun (ty, env_extension) -> Is_int ty, env_extension)
-      | n1 :: n2 :: [] ->
-        (* Note: Set.elements returns a sorted list, so n1 = 1 && n2 = 0
-           should never occur *)
-        if I.equal n1 I.zero && I.equal n2 I.one
-        then Ok (Is_int ty, TEE.empty ())
-        else bad_meet_or_join env t1 t2
-      | _::_::_::_ -> bad_meet_or_join env t1 t2
+        begin match shape with
+        | Some shape ->
+          Or_bottom.map
+            (T.meet env ty shape)
+            ~f:(fun (ty, env_extension) -> Is_int ty, env_extension)
+        | None -> Bottom
+        end
+      | _ :: _ :: _ ->
+        (* Note: we're potentially losing precision because the set could end up
+           not containing either 0 or 1 or both, but this should be uncommon. *)
+        Ok (Is_int ty, TEE.empty ())
       end
     | Naked_immediates is_int, Is_int ty ->
       begin match I.Set.elements is_int with
```

Comment:
I think this case can be shared with the previous one using an or-pattern now.

Reply:
I've left it as is because of the call to `T.meet`. I think it is the case that `T.meet env t1 t2` and `T.meet env t2 t1` are equivalent, but I wasn't completely sure.